crates/librqbit/src/blocklist.rs

```diff
@@ -0,0 +1,314 @@
+    }
+
+    pub async fn load_from_file(path: &str) -> Result<Self> {
+        let file = tokio::fs::File::open(path).await?;
+        let reader = tokio::io::BufReader::new(file);
+        Self::create_from_stream(reader).await
+    }
+
+    async fn create_from_stream<R>(reader: R) -> Result<Self>
+    where
+        R: AsyncRead + Unpin + Send,
+    {
+        let mut peek_bytes = [0u8; 2];
+        let mut reader = tokio::io::BufReader::new(reader);
+
+        // Peek the first bytes by filling buffer
+        let buffer = reader.fill_buf().await?;
+        if buffer.len() >= 2 {
+            peek_bytes.copy_from_slice(&buffer[0..2]);
+        } else {
+            return Err(anyhow::anyhow!(
+                "Content too short: not enough data to determine compression"
+            ));
+        }
+
+        // Check for Gzip magic bytes (1F 8B)
+        let is_gzip = peek_bytes == [0x1F, 0x8B];
+
+        let reader: Pin<Box<dyn AsyncRead + Send>> = if is_gzip {
+            trace!("Detected Gzip file, decompressing...");
+            Box::pin(BufReader::new(GzipDecoder::new(reader)))
+        } else {
+            trace!("Plain text file detected.");
+            Box::pin(reader)
+        };
+
+        let reader = BufReader::new(reader);
+        let mut lines = reader.lines();
+        let mut ip_ranges: Vec<std::ops::Range<IpAddr>> = Vec::new();
+        while let Some(line) = lines.next_line().await? {
+            // Skip comments and empty lines
+            if line.starts_with('#') || line.trim().is_empty() {
+                continue;
+            }
+
+            if let Some((start_ip, end_ip)) = parse_ip_range(&line) {
+                let range = start_ip..(increment_ip(end_ip).unwrap());
+                ip_ranges.push(range);
+            }
+        }
+
+        info!(
+            ip_entry_count = ip_ranges.len(),
+            "Finished loading blocklist"
+        );
+
+        let blocklist = Self::new(ip_ranges);
+        Ok(blocklist)
+    }
+
+    pub fn is_blocked(&self, ip: &IpAddr) -> bool {
+        self.blocked_ranges.query_point(*ip).next().is_some()
+    }
+}
+
+/// Safely increments an `IpAddr`, returning `None` if it would overflow.
+fn increment_ip(ip: IpAddr) -> Option<IpAddr> {
+    match ip {
+        IpAddr::V4(ipv4) => {
+            let num = u32::from_be_bytes(ipv4.octets());
+            num.checked_add(1).map(|n| IpAddr::V4(Ipv4Addr::from(n)))
+        }
+        IpAddr::V6(ipv6) => {
+            let num = u128::from_be_bytes(ipv6.octets());
+            num.checked_add(1).map(|n| IpAddr::V6(Ipv6Addr::from(n)))
+        }
+    }
+}
+
+fn parse_ip_range(line: &str) -> Option<(IpAddr, IpAddr)> {
+    // Skip comments and empty lines
+    let line = line.trim();
+    if line.starts_with('#') || line.is_empty() {
+        return None;
+    }
+
+    let is_ipv4 = line.matches('.').count() >= 6;
+    // Find the split point based on whether it's IPv4 or not
+    let split_point: usize = if is_ipv4 {
```

Comment:
why is this different for ipv4 and ipv6?

why not just

```
if let Some(rule_name, range) = split_once(':') {
    if let Some(start, end) = range.split_once('-') {
        ...
    }
}
```

Reply:
While testing against some real world block list i found out that some rules contained colon ()':') in the rule name.
This would cause the rule to fail parsing. 
I think we could safely ignore those rules though since they are not well formatted.